support OAuth

Nimrod007 opened the pull request.

Commit:
support OAuth

Nimrod007 commented:
WIP
@ohadraz please take a look to see if it fits 

it still isnt passing since i want to create some test account i can put its keys into source code

ohadraz commented on libpay-braintree/src/main/scala/com/wix/pay/braintree/BraintreeMerchant.scala:
```diff
@@ -1,6 +1,15 @@
+ * @param merchantAccountIds  Maps currency (ISO 4217) to merchant account id. on absence default is used (can be set in Braintree account)
  */
-case class BraintreeMerchant(merchantId: String, publicKey: String, privateKey: String, merchantAccountIds: Map[String, String])
+class BraintreeMerchantBase(val merchantAccountIds: Option[Map[String, String]])
```
Consider renaming it to "BraintreeMerchant".
Also, what about making this somewhat of a marker interface?
Something like:
trait BraintreeMerchant {
   def merchantAccountIds: Option[Map[String, String]] = None
}

ohadraz commented on libpay-braintree/src/main/scala/com/wix/pay/braintree/BraintreeMerchant.scala:
```diff
@@ -1,6 +1,15 @@
-case class BraintreeMerchant(merchantId: String, publicKey: String, privateKey: String, merchantAccountIds: Map[String, String])
+class BraintreeMerchantBase(val merchantAccountIds: Option[Map[String, String]])
+
+case class BraintreeMerchant(merchantId: String,
```
If you accept the renaming of "BraintreeMerchantBase" to "BraintreeMerchant", then this should be renamed to avoid naming collision. I am pretty bad with naming, but, maybe "BraintreeMerchantLegacy", or "SimpleBraintreeMerchant", or "CredentialsBasedBraintreeMerchant"?

Nimrod007 replied:
im not good at names but CredentialsBasedBraintreeMerchant and OAuthBraintreeMerchant sounds good

ohadraz commented on libpay-braintree/src/main/scala/com/wix/pay/braintree/BraintreeMerchantParser.scala:
```diff
@@ -1,6 +1,6 @@
 package com.wix.pay.braintree
 
 trait BraintreeMerchantParser {
```
How about making this type parameterized, like :
trait BraintreeMerchantParser[+T <: BraintreeMarchant] { 
   def parse(merchantKey: String): T
  def stringify(merchant: T): String
}
(this assuming you accept my previous comment, regarding renaming BraintreeMerchantBase to BraintreeMerchant)

Nimrod007 replied:
not a big fan of parameterized since its only 2 merchant types

ohadraz replied:
Hmm.... Remember what I said when you talked to me about this PR? That I believe one parser should handle both cases? This just maybe the way to accomplish that.
Would you consider to re-think it?

Nimrod007 replied:
ok so i moved to 1 merchant parser

ohadraz commented on libpay-braintree/src/main/scala/com/wix/pay/braintree/JsonBraintreeOAuthMerchantParser.scala:
```diff
@@ -0,0 +1,16 @@
+import org.json4s.DefaultFormats
+import org.json4s.native.Serialization
+
+class JsonBraintreeOAuthMerchantParser() extends BraintreeMerchantParser {
```
If accept my comment regarding type parameterized parser trait, you can check if this class is required altogether.
I'm not sure about what I'm about to suggest, with erasure and all, but give it a try:
If you change the current implementation of BraintreeMerchantParser to something like:
class JsonBraintreeMerchantParser[T] extends BraintreeMerchantParser[T] {
   override def parse(merchantKey: String): T = {
      Serialization.read[T](merchantKey)
   }

   override def stringify(merchant: T): String = {
      Serialization.write(merchant)
   }
}

Then this class may not be required.

Nimrod007 replied:
@dleshem has just jumped behind me and suggested that parser will know which merchant it is (hints) so we dont need 2 parsers (so we wont need 2 gateways) 

will do that and update PR

Nimrod007 commented on libpay-braintree/src/test/scala/com/wix/pay/braintree/it/BraintreeGatewayWithOAuthIT.scala:
```diff
@@ -0,0 +1,46 @@
+import org.specs2.mutable.SpecWithJUnit
+
+
+class BraintreeGatewayWithOAuthIT extends SpecWithJUnit {
```
@ohadraz - are you guys ok that the test will use Braintree Sandbox Env ? i think its valid, if its starting failing due to networking i can add driver / wiremock

ohadraz replied:
I do not think this would be a good idea.
Remember that this is an open source project; I do not think we want to put Wix credentials to Braintree Sandbox.
And regardless, and as Danny mentioned, as an open source, this should be the state of the art of Wix engineering. Using the sandbox is not a good example from TDD, which we preach on at Wix.

Nimrod007 replied:
ok will use wire mock (in serious TDD you test against Sandbox minimal set, and extended scenarios with a driver. if its stable you win it all, if its not you move just to the driver)

ohadraz replied:
Not sure I agree. You may do that as "learning tests", which are later removed.
Or, have tests against Sandbox in a different project, that runs daily/weekly/monthly/yearly/millennially. Failures in this project alerts you that something may have gone wrong, or that there just had been a network issue or downtime on the Sandbox end, you will have to check, but it does not break your "real" project...

Nimrod007 replied:
added a driver :)

Commit:
move to 1 merchant parser

add IT test for OAuth merchant

Nimrod007 commented:
ok so i moved to 1 merchant parser to parse them all (so you can create 1 payment gateway)

added a driver to the new OAuth flow (using wiremock)

tell me what you think

ohadraz reviewed (commented):
I think it looks much better :)
Two things, though:
One, is the base merchant - I saw you left it as a class, rather than my suggestion of making it an interface, perhaps with default implementation (merchant accounts equal to None). With Scala 2.12, which runs with Java 1.8, this should be used as an interface with default method.
The second thing is the token in the driver. It is not a real token, right?...

Nimrod007 commented:
regarding 1 np,
token is fake
